### ISSUE TITLE
Cannot reproduce the 64.1 mAP on COCO dataset by yolov5m

### ISSUE THREAD

**Issue by SikaAntler** [ultralytics/yolov5#10905]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I was trying to reproduce the **64.1** mAP(0.5) by yolov5m on COCO dataset. 

My environment is:
`Python-3.9.13 torch-1.9.1+cu111`

My training shell script is:
`python -m torch.distributed.run --nproc_per_node 4 train.py --weights '' --cfg yolov5m.yaml --data data/coco.yaml --hyp data/hyps/hyp.scratch-high.yaml --epochs 300 --batch-size 128 --imgsz 640 --device 0,1,2,3 --sync-bn`

Then, I could only get the best mAP(0.5) of **0.587**. 

I have been mentioned by other issue that the mAP was reported by pycocotools, not the inside metrics. Then, I download the support chechpoint of yolov5m to test by the inside metrics, then I got the mAP(0.5) of **0.635**. There is still a big gap between **0.587** and **0.635**.

So I considered that the problem might be caused by the training process, can anyboby help me to solve this? Thank you very much!

**Comment by marigoold** [comment]
I encountered the same problem, but it's on YOLOv5n, I can't reproduce the 45.7 mAP :( YOLOv5 early stopped at ~36 mAP

**Comment by SikaAntler** [comment]
> I encountered the same problem, but it's on YOLOv5n, I can't reproduce the 45.7 mAP :( YOLOv5 early stopped at ~36 mAP

Ok, then I reproduce the yolov5n, and I will tell you my result tomorrow.

**Comment by marigoold** [comment]
> > I encountered the same problem, but it's on YOLOv5n, I can't reproduce the 45.7 mAP :( YOLOv5 early stopped at ~36 mAP
> 
> Ok, then I reproduce the yolov5n, and I will tell you my result tomorrow.

Thanks! BTW, my training command is `python3 -m torch.distributed.launch --nproc_per_node 8 train.py --data coco.yaml --epochs 300 --weights '' --cfg yolov5n.yaml  --batch-size 512 --sync-bn`, and used the main branch, FYI

**Comment by marigoold** [comment]
And I disabled amp training for loss computing, and the mAP seems to be normal (I didn't finish the training, just check the increasing trend of mAP). I tried different training combinations, the results are:
- DDP and AMP enabled, got abnormal mAP
- DDP enabled, AMP disabled(comment the amp related code in train.py), got normal mAP
- DDP disabled, AMP enabled (single GPU training), got normal mAP

**Comment by SikaAntler** [comment]
> > > I encountered the same problem, but it's on YOLOv5n, I can't reproduce the 45.7 mAP :( YOLOv5 early stopped at ~36 mAP
> > 
> > 
> > Ok, then I reproduce the yolov5n, and I will tell you my result tomorrow.
> 
> Thanks! BTW, my training command is `python3 -m torch.distributed.launch --nproc_per_node 8 train.py --data coco.yaml --epochs 300 --weights '' --cfg yolov5n.yaml --batch-size 512 --sync-bn`, and used the main branch, FYI

Well, I could only get **0.32** mAP(0.5), and no early stop, I don't know why...

My training script is:
`python -m torch.distributed.run --nproc_per_node 4 train.py --weights '' --cfg yolov5n.yaml --data coco.yaml --hyp hyp.scratch-low.yaml --epochs 300 --batch-size 512 --imgsz 640 --device 0,1,2,3 --sync-bn`

**Comment by SikaAntler** [comment]
> And I disabled amp training for loss computing, and the mAP seems to be normal (I didn't finish the training, just check the increasing trend of mAP). I tried different training combinations, the results are:
> 
> * DDP and AMP enabled, got abnormal mAP
> * DDP enabled, AMP disabled(comment the amp related code in train.py), got normal mAP
> * DDP disabled, AMP enabled (single GPU training), got normal mAP

Thank you, it seems like the problem caused by DDP and AMP. I am going to try these settings.

**Comment by marigoold** [comment]
> > And I disabled amp training for loss computing, and the mAP seems to be normal (I didn't finish the training, just check the increasing trend of mAP). I tried different training combinations, the results are:
> > 
> > * DDP and AMP enabled, got abnormal mAP
> > * DDP enabled, AMP disabled(comment the amp related code in train.py), got normal mAP
> > * DDP disabled, AMP enabled (single GPU training), got normal mAP
> 
> Thank you, it seems like the problem caused by DDP and AMP. I am going to try these settings.

Unfortunately, DDP disabled, AMP enabled (single GPU training) seems to get abnormal mAP too..
I just trained YOLOv5n on single gpu, and mAP(0.5) stuck at ~0.354.
Maybe the issue is caused by AMP, and has no business with DDP

**Comment by YoungjaeDev** [comment]
@marigoold @SikaAntler 
When I turn to ddp, for example,
```bash
 python -m torch.distributed.run --nproc_per_node 4 train.py --batch 256 --cfg models/yolov5s.yaml --weights '' --data coco.yaml --img 640 --epochs 300 --project Train_COCO --name coco-scratch-v5s640 --device 0,1,2,3 --hyp data/hyps/hyp.scratch-low.yaml
```
Is the `sync-bn` option required?  can use both.

**Comment by marigoold** [comment]
> @marigoold @SikaAntler When I turn to ddp, for example,
> 
> ```shell
>  python -m torch.distributed.run --nproc_per_node 4 train.py --batch 256 --cfg models/yolov5s.yaml --weights '' --data coco.yaml --img 640 --epochs 300 --project Train_COCO --name coco-scratch-v5s640 --device 0,1,2,3 --hyp data/hyps/hyp.scratch-low.yaml
> ```
> 
> Is the `sync-bn` option required? can use both.

Hi, @youngjae-avikus you can read this [discussion](https://discuss.pytorch.org/t/ddp-sync-bn-always/104783). Briefly, sync-bn will improve your accuracy for BN, but slow down the training.

**Comment by SikaAntler** [comment]
> @marigoold @SikaAntler When I turn to ddp, for example,
> 
> ```shell
>  python -m torch.distributed.run --nproc_per_node 4 train.py --batch 256 --cfg models/yolov5s.yaml --weights '' --data coco.yaml --img 640 --epochs 300 --project Train_COCO --name coco-scratch-v5s640 --device 0,1,2,3 --hyp data/hyps/hyp.scratch-low.yaml
> ```
> 
> Is the `sync-bn` option required? can use both.

In my experience, `sync-bn` is not necessary, you can drop it when your batch-size is big enough in single gpu.

**Comment by SikaAntler** [comment]
> > > And I disabled amp training for loss computing, and the mAP seems to be normal (I didn't finish the training, just check the increasing trend of mAP). I tried different training combinations, the results are:
> > > 
> > > * DDP and AMP enabled, got abnormal mAP
> > > * DDP enabled, AMP disabled(comment the amp related code in train.py), got normal mAP
> > > * DDP disabled, AMP enabled (single GPU training), got normal mAP
> > 
> > 
> > Thank you, it seems like the problem caused by DDP and AMP. I am going to try these settings.
> 
> Unfortunately, DDP disabled, AMP enabled (single GPU training) seems to get abnormal mAP too.. I just trained YOLOv5n on single gpu, and mAP(0.5) stuck at ~0.354. Maybe the issue is caused by AMP, and has no business with DDP

I have tried yolov5m with AMP disabled, however, the best mAP(0.5) only achieved **0.58484**.

My method of turning off AMP is to modify `amp = check_amp(model)` always be `amp = False`.  Am I correct?

**Comment by glenn-jocher** [comment]
Use batch size 128 to reproduce official trainings

**Comment by marigoold** [comment]
> Use batch size 128 to reproduce official trainings

@glenn-jocher Hi, how to reproduce official mAP on multi cards? It seems the scripts in https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training always get low mAP : (

**Comment by SikaAntler** [comment]
> Use batch size 128 to reproduce official trainings

Yes, I set the batch-size to 128 and turn on `sync-bn`, however, I could only get the best mAP(0.5) of **0.58691**.

As reminded by marigold, I turned off AMP, the best mAP(0.5) is only **0.58484**.

Is there any key point I have missed? Thank you.

**Comment by eypros** [comment]
Hello, a bit late but a key factor you are changing is using image size 512 for some reason. The reported results uses 640 for yolov5n.pt and 1280x1280 for yolov5n6.pt as I get it. You should check with this image size for sure to make clear conclusions.

**Comment by glenn-jocher** [comment]
Yes, @eypros is correct. The original YOLOv5 research paper and the YOLOv5 github repository both using 640x640 image size for the COCO dataset. 

If you use different image sizes than the original research, the performance of the model will be different. Hence, I suggest using the original image size for accurate results.
